docs/features/routing.rst

```diff
@@ -235,4 +235,36 @@ Ocelot will also allow you to put query string parameters in the UpstreamPathTem
     }
 
 In this example Ocelot will only match requests that have a matching url path and the query string starts with unitId=something. You can have other queries after this
-but you must start with the matching parameter. Also Ocelot will swap the {unitId} parameter from the query string and use it in the downstream request path. 
+but you must start with the matching parameter. Also Ocelot will swap the {unitId} parameter from the query string and use it in the downstream request path.
+
+Upstream header-based routing
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This feature was requested in `issue 360 <https://github.com/ThreeMammals/Ocelot/issues/360>`_ and `issue 624 <https://github.com/ThreeMammals/Ocelot/issues/624>`_.
+
+Ocelot allows you to define a Route with upstream headers, each of which may define a set of accepted values. If a Route has a set of upstream headers defined in it, it will no longer match a request's upstream path based solely on upstream path template. The request must also contain one or more headers required by the Route for a match.
+
+A sample configuration might look like the following:
+
+.. code-block:: json
+
+    {
+        "Routes": [
+            {
+                "UpstreamHeaderRoutingOptions": {
+                    "Headers": {
+                        "X-API-Version": [ "1", "2" ],
+                        "X-Tennant-Id": [ "tennantId" ]
+                    },
+                    "TriggerOn": "all"
+                }
+            }
+        ]
+    }

```

Comment:
Fair enough, I'll add the other parts of the `Route` example.

Reply:
Did you add these parts already? What's commit hash?
Please, don't comment if there is nothing to show, nothing fixed.

Reply:
Still working on the fixes. I was sick for the past week or so.

Reply:
Having easy recovery!
Don't drink cold beer :beer:  and don't eat ice cream :icecream:  while walking through the hot streets of Prague! :wink:

Reply:
Perfect! I see some changes in https://github.com/ThreeMammals/Ocelot/pull/1684/commits/6d29040db6570a01847ef596a2cdc583e146ad6c after [line 253](commits/6d29040db6570a01847ef596a2cdc583e146ad6c#diff-f8cdcb9e36448707f046f3a7ee0f861135cef70cd2fe0153a82a0873bf68c45eR253), that's good!
But now readability has become worse!
Well... The reason of raising of this issue is "inability to recognize changes between classic route and header-based one".
So, it is not clear what are the changes in URLs. **We need to focus on this problem.**
I suggest to:
- emphasize by comments in JSON key changes (new properties) defining block by comments (`// begin` and `// end`)
- before JSON sample ([lines 251-291](/mrclayman/Ocelot/blob/upstream-header-routing-2023/docs/features/routing.rst?plain=1#L251-L291)) or after, explain to user/developer how these setup settings change entire URLs of upstream and downstream

Reply:
I've changed my opinion.
Because there is no sense to show the values of **Downstream*** and **Upstream*** properties, it is better to remove them the the code block. So, explaining **UpstreamHeaderRoutingOptions** property is sufficient.

✔️ Fixed in https://github.com/ThreeMammals/Ocelot/pull/1684/commits/44df12072bae0a9e46f6ba20ffdaa3c8b82bf97a